### ISSUE TITLE
Query: Error using a compare with a nullable boolean

### ISSUE THREAD

**Issue by lucamorelli** [aspnet/EntityFramework#5877]
Hi, 
  I just updated to asp.net core rtm an application and I'm having a runtime error
  I have this expression:

```
clienti = db.Clienti
   .Where(cli => cli.Attivo && cli.Operativo != null && cli.Operativo.Value)
   .OrderBy(cli => cli.Descrizione)
   .ToList();

```
where Attivo is a bool and Operativo is a ?bool.
If I run this code I obtain this error message

```
Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryCompilationContextFactory:Error: An exception occurred in the database while iterating the results of a query.
System.InvalidOperationException: The binary operator Equal is not defined for the types 'System.Nullable`1[System.Boolean]' and 'System.Boolean'.

```
Has anything changed ? This code worked fine before the upgrade.

**Comment by gdoron** [comment]
Regardless of the error, I would have written the query with `??`:

	clienti = db.Clienti
	   .Where(cli => cli.Attivo && cli.Operativo ?? false)
	   .OrderBy(cli => cli.Descrizione)
	   .ToList();

**Comment by lucamorelli** [comment]
I never knew I could use ?? inside a linq expression, anyway I don't like it very much and not very clear

**Comment by gdoron** [comment]
@lucamorelli that's your call, but semantically, this is the right way IMHO of doing it since it's the equivalent of `ISNULL(Operativo, 0) = 1`

Anyway, I remember I saw already an issue regarding cases where you need to "help" EF with nullable stuff.

If I remember correctly the solution was casting to `bool?`

**Comment by rowanmiller** [comment]
It should also be possible to write this as `cli.Operativo == true` - which would evaluate to `false` when `Operativo` is `null` or `false`. This was reported in https://github.com/aspnet/EntityFramework/issues/5899 and should be addressed as part of the fix to this issue too

**Comment by rowanmiller** [comment]
Also make sure https://github.com/aspnet/EntityFramework/issues/5935 is addressed.

**Comment by bigteejay** [comment]
My similar (not same?) repo of what is triggering the same exception after my upgrade to core...

```
var cnt = _db.RegisteredUsers
					.Where(r => 
						r.UserId == response.UserId && 
						(
							(r.Active.HasValue && r.Active.Value) ||
							(r.Admin.HasValue && r.Admin.Value)
						)
					)
					.Count();
```

I thought checking `.HasValue` before checking for `true` would save me from this, apparently not?  I can't currently get the SQL statement it is attempting (just the exception logged to the console.)

if it matters, in `RegisteredUsers`, the first two records have `null` values for both `Active` and `Admin` for the `UserId` but the final record for that user has values in those fields.

**Comment by bigteejay** [comment]
I tried rewriting the where clause with Rowan's [comment](https://github.com/aspnet/EntityFramework/issues/5877#issuecomment-229506242) in mind...

```
(r.Active.Value == true) ||
(r.Admin == true)
```

I figured that top line would fail for the same reasons trying `.HasValue` and then `.Value` properties are (implied?) to be true in my original code.

I figured the second line would fail because `.Admin` is in fact of type `bool?` and I thought the compiler issued errors/warnings when comparing a `bool` (`true') to a `bool?`

**But**, to my surprise, both worked!?

**Comment by bigteejay** [comment]
I got my code working using @gdoron's [suggestion](https://github.com/aspnet/EntityFramework/issues/5877#issuecomment-228872924) of using `?? false` to test true with nullable properties.

**Comment by smitpatel** [comment]
merged to 1.0.1 via add3939bd592f1c4a7f249b81d024a2ff260091d

**Comment by ruant** [comment]
Did this make it into the 1.0.1 release?
I still have issues with this.

```
var currentProfile = _db.CustomerProfileSet
  .Where(customerProfile => customerProfile.CustomerId == id && customerProfile.Current == true);
```

This wil fail with:
```
System.InvalidOperationException: The binary operator Equal is not defined for the types 'System.Nullable`1[System.Boolean]' and 'System.Boolean'
```

I fixed it with changing the code to:
```
var currentProfile = _db.CustomerProfileSet
  .Where(customerProfile => customerProfile.CustomerId == id && customerProfile.Current.Equals(true));
```

**Comment by rowanmiller** [comment]
Yes, the fix for this issue was included in 1.0.1. Please open a new issue with the problem you are seeing.
